cmd/license-update.go

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2015-2023 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/minio/minio/internal/logger"
+	"github.com/tidwall/gjson"
+)
+
+const (
+	licUpdateCycle = 24 * time.Hour * 30
+	licRenewURL    = "https://subnet.min.io/api/cluster/renew-license"
+	licRenewURLDev = "http://localhost:9000/api/cluster/renew-license"
+)
+
+var licUpdateLeaderLockTimeout = newDynamicTimeout(30*time.Second, 10*time.Second)
+
+// initlicenseUpdateJob start the periodic license update job in the background.
+func initLicenseUpdateJob(ctx context.Context, objAPI ObjectLayer) {
+	go func() {
+		r := rand.New(rand.NewSource(time.Now().UnixNano()))
+		// Leader node (that successfully acquires the lock inside runLicenseUpdate)
+		// will keep performing the license update. If the leader goes down for some
+		// reason, the lock will be released and another node will acquire it and
+		// take over because of this loop.
+		for {
+			runLicenseUpdate(ctx, objAPI)
+
+			// license update running on a different node.
+			// sleep for some time and try again.
+			duration := time.Duration(r.Float64() * float64(licUpdateCycle))
+			if duration < time.Hour {
+				// Make sure to sleep atleast an hour.
+				duration = time.Hour
+			}
+			time.Sleep(duration)
+		}
+	}()
+}
+
+func runLicenseUpdate(ctx context.Context, objAPI ObjectLayer) {
+	// Make sure only 1 license update job is running on the cluster.
+	locker := objAPI.NewNSLock(minioMetaBucket, "licenseUpdate/runLicenseUpdate.lock")
```

Comment:
@vadmeste can you please help me with this? I've kept this similar to other periodic jobs like callhome.

Reply:
> @vadmeste can you please help me with this? I've kept this similar to other periodic jobs like callhome.

@anjalshireesh, This will be done by one node in a cluster, right?

you can do this way in the beginning in this function, runLicenseUpdate()
```
        ctx, cancel := globalLeaderLock.GetLock(ctx)
        defer cancel()
```

I am assuming runLicenseUpdate() is infinite and will never stop unless the server is stopped/killed.

By the way, can you rename runLicenseUpdate() to licenceUpdaterLoop() ?

Reply:
Thanks @vadmeste , will do :+1:

Reply:
@klauspost @vadmeste have implemented the suggested changes. PTAL.